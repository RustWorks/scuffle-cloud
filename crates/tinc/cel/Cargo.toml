[package]
name = "tinc-cel"
version = "0.1.0"
edition = "2024"
description = "A CEL Type used by tinc-build and tinc for validating protobuf messages"
repository = "https://github.com/scufflecloud/scuffle"
authors = ["Scuffle <opensource@scuffle.cloud>"]
readme = "README.md"
documentation = "https://docs.rs/tinc-build"
license = "MIT OR Apache-2.0"
keywords = ["grpc", "protobuf", "tonic", "codegen"]

[dependencies]
num-traits = "0.2.19"
bytes = "1.5"
float-cmp = "0.10"
thiserror = "2"
chrono = "0.4"
regex = "1"
linkme = "0.3"
fmtools = "0.1"
uuid = "1"
url = "2"
email_address = "0.2.9"
scuffle-workspace-hack.workspace = true

[features]
runtime = []

