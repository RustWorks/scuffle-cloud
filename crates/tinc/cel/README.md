<!-- cargo-sync-rdme title [[ -->
# tinc-cel
<!-- cargo-sync-rdme ]] -->

> [!WARNING]  
> This crate is under active development and may not be stable.

<!-- cargo-sync-rdme badge [[ -->
![License: MIT OR Apache-2.0](https://img.shields.io/crates/l/tinc-cel.svg?style=flat-square)
[![docs.rs](https://img.shields.io/docsrs/tinc-cel.svg?logo=docs.rs&style=flat-square)](https://docs.rs/tinc-cel)
[![crates.io](https://img.shields.io/crates/v/tinc-cel.svg?logo=rust&style=flat-square)](https://crates.io/crates/tinc-cel)
[![GitHub Actions: ci](https://img.shields.io/github/actions/workflow/status/scufflecloud/scuffle/ci.yaml.svg?label=ci&logo=github&style=flat-square)](https://github.com/scufflecloud/scuffle/actions/workflows/ci.yaml)
[![Codecov](https://img.shields.io/codecov/c/github/scufflecloud/scuffle.svg?label=codecov&logo=codecov&style=flat-square)](https://codecov.io/gh/scufflecloud/scuffle)
<!-- cargo-sync-rdme ]] -->

---

<!-- cargo-sync-rdme rustdoc [[ -->
Currently this is a fully private api used by `tinc` and `tinc-build` to
compile and execute [CEL](https://cel.dev/) expressions.
Check out the [changelog](./CHANGELOG.md).

### Feature flags

* **`runtime`** —  Enables runtime enum type indexing
* **`docs`** —  Enables changelog and documentation of feature flags

### License

This project is licensed under the MIT or Apache-2.0 license.
You can choose between one of them if you use this work.

`SPDX-License-Identifier: MIT OR Apache-2.0`
<!-- cargo-sync-rdme ]] -->
