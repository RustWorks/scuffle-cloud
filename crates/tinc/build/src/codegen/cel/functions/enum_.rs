use syn::parse_quote;
use tinc_cel::CelValue;

use super::Function;
use crate::codegen::cel::compiler::{CompileError, CompiledExpr, CompilerCtx, ConstantCompiledExpr, RuntimeCompiledExpr};
use crate::codegen::cel::types::CelType;
use crate::types::{ProtoModifiedValueType, ProtoPath, ProtoType, ProtoValueType};

#[derive(Debug, Clone, Default)]
pub(crate) struct Enum(pub Option<ProtoPath>);

impl Function for Enum {
    fn name(&self) -> &'static str {
        "enum"
    }

    fn syntax(&self) -> &'static str {
        "<this>.enum() | <this>.enum(<path>)"
    }

    fn compile(&self, ctx: CompilerCtx) -> Result<CompiledExpr, CompileError> {
        let Some(this) = ctx.this.as_ref() else {
            return Err(CompileError::syntax("missing this", self));
        };

        if ctx.args.len() > 1 {
            return Err(CompileError::syntax("invalid number of arguments", self));
        }

        let enum_path = if let Some(arg) = ctx.args.first() {
            ctx.resolve(arg)?
        } else {
            match (&this, &self.0) {
                (
                    CompiledExpr::Runtime(RuntimeCompiledExpr {
                        ty:
                            CelType::Proto(ProtoType::Modified(ProtoModifiedValueType::Optional(ProtoValueType::Enum(path))))
                            | CelType::Proto(ProtoType::Value(ProtoValueType::Enum(path))),
                        ..
                    }),
                    _,
                )
                | (_, Some(path)) => CompiledExpr::Constant(ConstantCompiledExpr {
                    value: CelValue::String(path.0.clone().into()),
                }),
                _ => {
                    return Err(CompileError::syntax(
                        "unable to determine enum type, try providing an explicit path",
                        self,
                    ));
                }
            }
        };

        let this = this.clone().into_cel()?;
        let enum_path = enum_path.into_cel()?;

        match (this, enum_path) {
            (
                CompiledExpr::Constant(ConstantCompiledExpr { value: this }),
                CompiledExpr::Constant(ConstantCompiledExpr { value: enum_path }),
            ) => Ok(CompiledExpr::constant(CelValue::cel_to_enum(this, enum_path)?)),
            (this, enum_path) => Ok(CompiledExpr::runtime(
                CelType::CelValue,
                parse_quote! {
                    ::tinc::__private::cel::CelValue::cel_to_enum(
                        #this,
                        #enum_path,
                    )?
                },
            )),
        }
    }
}
