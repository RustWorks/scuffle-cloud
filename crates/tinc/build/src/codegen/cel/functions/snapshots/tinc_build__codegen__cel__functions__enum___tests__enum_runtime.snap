---
source: crates/tinc/build/src/codegen/cel/functions/enum_.rs
expression: result
---
Runtime(
    RuntimeCompiledExpr {
        ty: CelValue,
        expr: ::tinc::__private::cel::CelValue::cel_to_enum(
            ::tinc::__private::cel::CelValueConv::conv(input),
            ::tinc::__private::cel::CelValue::String(
                ::tinc::__private::cel::CelString::Borrowed("some.Enum"),
            ),
        )?,
    },
)
